### PR TITLE
Azure: Add node management

### PR DESCRIPTION
**Pull request by alvaroaleman**
This makes the neccesary changes to the hostedcluster and nodepool
controller as well as to the controlplane operator and hosted cluster
config operator to allow creation of azure nodes to succeed. These nodes
join the cluster and become ready. What is not part of this and will be
done as a follow-up is:
* DNS setup
* All Clusteroperators become ready
* Setting up the in-tree cloudprovider

Ref https://issues.redhat.com/browse/HOSTEDCP-242

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.
-

/cc @csrwng @enxebre

### CODE REVIEW

File: cmd/infra/azure/create.go [hunk 1]
```diff
@@ -92,6 +92,7 @@ type CreateInfraOutput struct {
 	ResourceGroupName string `json:"resourceGroupName"`
 	VNetID            string `json:"vnetID"`
 	VnetName          string `json:"vnetName"`
+	SubnetNamge       string `json:"subnetName"`
```
Comment (csrwng) [inline cmd/infra/azure/create.go:95]: s/SubnetNamge/SubnetName/

**Comment by csrwng** [comment]
Other than the nit above, this lgtm. 
@alvaroaleman do you know if there's something equivalent to STS in azure such that we don't have to hold customer creds?

File: cmd/infra/azure/create.go [hunk 2]
```diff
@@ -223,6 +224,10 @@ func (o *CreateInfraOptions) Run(ctx context.Context) (*CreateInfraOutput, error
 	if err != nil {
 		return nil, fmt.Errorf("failed to get vnet creation result: %w", err)
 	}
+	if vnet.Subnets == nil || len(*vnet.Subnets) < 1 {
+		return nil, fmt.Errorf("created vnet has no subnets: %+v", vnet)
+	}
+	result.SubnetNamge = *(*vnet.Subnets)[0].Name
```
Comment (enxebre) [inline cmd/infra/azure/create.go:230]: typo `subnetName`

File: hypershift-operator/controllers/hostedcluster/internal/platform/azure/azure.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/openshift/hypershift/support/upsert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capiazure "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/internal/platform/azure/azure.go:17]: Let's pick image with your changes from https://console.cloud.google.com/gcr/images/k8s-staging-cluster-api-azure/global/cluster-api-azure-controller

File: hypershift-operator/controllers/hostedcluster/internal/platform/platform.go [hunk 4]
```diff
@@ -27,11 +28,7 @@ type Platform interface {
 	// Implementations should use the given input and client to create and update the desired state of the
 	// platform infrastructure CAPI CR, which will then be referenced by the CAPI Cluster CR.
 	// TODO (alberto): Pass createOrUpdate construct instead of client.
-	ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN,
-		hcluster *hyperv1.HostedCluster,
-		controlPlaneNamespace string,
-		apiEndpoint hyperv1.APIEndpoint,
-	) (client.Object, error)
+	ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN, hcluster *hyperv1.HostedCluster, controlPlaneNamespace string, apiEndpoint hyperv1.APIEndpoint) (client.Object, error)
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/internal/platform/platform.go:31]: this is not changing the signature but just inlining everything, right?
fwiw I broke it up this way so lines beyond ~110 get difficult to read in 13-inch. I'm fine with the change.

File: hypershift-operator/controllers/nodepool/azure.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	"encoding/base64"
+	"fmt"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"golang.org/x/crypto/ssh"
+	capiazure "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+func azureMachineTemplateSpec(hcluster *hyperv1.HostedCluster, nodePool *hyperv1.NodePool, existing capiazure.AzureMachineTemplateSpec) (*capiazure.AzureMachineTemplateSpec, error) {
+	// The azure api requires to pass a public key. This key is randomly generated, the private portion is thrown away and the public key
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/azure.go:15]: :/ is there a link to azure docs we can drop here?

File: api/v1alpha1/hostedcluster_types.go [hunk 6]
```diff
@@ -637,6 +637,7 @@ type AzurePlatformSpec struct {
 	ResourceGroupName string                      `json:"resourceGroup"`
 	VnetName          string                      `json:"vnetName"`
 	VnetID            string                      `json:"vnetID"`
+	SubnetName        string                      `json:"subnetName"`
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:640]: why is this here instead of in NodePool API?

File: ignition-server/controllers/machineconfigserver_ignitionprovider.go [hunk 7]
```diff
@@ -183,6 +184,15 @@ func (p *MCSIgnitionProvider) GetPayload(ctx context.Context, releaseImage strin
 const mcoBootstrapScript = `#!/bin/bash
 mkdir -p /mcc-manifests/bootstrap/manifests
 mkdir -p /mcc-manifests/manifests
+{{ if .isAzure }}
+cat <<EOF >/tmp/cloud.conf.configmap.yaml
```
Comment (enxebre) [inline ignition-server/controllers/machineconfigserver_ignitionprovider.go:188]: why is this needed?

File: hypershift-operator/controllers/hostedcluster/internal/platform/platform.go (hunk 4 above)
Comment (alvaroaleman) [reply hypershift-operator/controllers/hostedcluster/internal/platform/platform.go:31]: Yeah it just changes the style, I did that because I find signatures that have multiple but not all parameters on a single line confusing to read, IMHO either all should be on one line like right now or each one gets their own line

File: api/v1alpha1/hostedcluster_types.go (hunk 6 above)
Comment (alvaroaleman) [reply api/v1alpha1/hostedcluster_types.go:637]: Because the assumption so far is that there is one subnet per cluster

File: ignition-server/controllers/machineconfigserver_ignitionprovider.go (hunk 7 above)
Comment (alvaroaleman) [reply ignition-server/controllers/machineconfigserver_ignitionprovider.go:188]: Because the kubelet panics if its started with the Azure cloud provider but no cloud config

**Comment by alvaroaleman** [comment]
> @alvaroaleman do you know if there's something equivalent to STS in azure such that we don't have to hold customer creds?

I don't think there is, unfortunately.

File: api/v1alpha1/hostedcluster_types.go (hunk 6 above)
Comment (csrwng) [reply api/v1alpha1/hostedcluster_types.go:637]: Is that the case with standalone OCP?

Comment (alvaroaleman) [reply api/v1alpha1/hostedcluster_types.go:637]: Standalone OCP creates two subnets, one for the masters, one for the workers: https://github.com/openshift/installer/blob/8fca1ade5b096d9b2cd312c4599881d099439288/data/data/azure/vnet/vnet.tf

But since we don't have masters, we don't need the former.

The [main reasons to create multiple subnets seem to be](https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-vnet-plan-design-arm#subnets):
a) Some azure resources create their own subnet
b) Security, network security groups are associated to a subnet

But since from an azure networking perspective all workers fall into the same security domain (they need to be able to talk to each other freely, reach the internet and be reachable by LBs) I don't see why we would need multiple. Cluster-level network security happens below azure through the clusters SDN.

Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go:637]: We should possibly expose the [AZ within the subnet](https://docs.microsoft.com/en-us/azure/availability-zones/az-overview#availability-zones) https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_types.go#L44.
This is also what we do in the [installer](https://github.com/openshift/installer/blob/master/pkg/asset/machines/worker.go#L360).

File: ignition-server/controllers/machineconfigserver_ignitionprovider.go (hunk 7 above)
Comment (enxebre) [reply ignition-server/controllers/machineconfigserver_ignitionprovider.go:188]: Is that only for azure or also for AWS?
Is there an issue open against kubelet azure provider code to discuss?
Can we unify code and do this unconditionally so no need for the ignition-server to know the platform?

File: api/v1alpha1/hostedcluster_types.go (hunk 6 above)
Comment (alvaroaleman) [reply api/v1alpha1/hostedcluster_types.go:637]: What do you mean by "AZ within the subnet"? The subnet is not scoped to an AZ, it is possible to put an instance in a different AZ but use the same subnet. There is a follow-up task for Azure multi-zone support: https://issues.redhat.com/browse/HOSTEDCP-322

Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go:637]: yeh I meant the list of az in that subnet, cool yeah so basically 332

**Comment by sjenning** [comment]
I think all feedback is address and it
/lgtm
